### PR TITLE
fix: only send exception responses if header is not already sent

### PR DESCRIPTION
**Pull request by wSedlacek**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## What is the new behavior?
When an exception is thrown it will trigger the exception filter to try to send a reply.
If a header was already sent then the exception filter will fail to send this reply.

## Does this PR introduce a breaking change?
- [x] Yes
- [ ] No

This change requires adapters to provide two new methods, `isHeaderSent()` and `end()`

### CODE REVIEW

File: packages/platform-express/adapters/express-adapter.ts [hunk 1]
```diff
@@ -60,6 +60,10 @@ export class ExpressAdapter extends AbstractHttpAdapter {
     return response.status(statusCode);
   }
 
+  public end(response: any, message?: string) {
+    return response.end(message !== undefined ? String(message) : undefined);
```
Comment (micalevisk) [inline packages/platform-express/adapters/express-adapter.ts:64]: I'm curious, why not just `response.end(message)`?

If `message` is `undefined`, then `response.end(undefined)` will work.  
If it's an string, then `String(message) === message`, ie.,  
`response.end(String(message))` end up having the same outcome of `response.end(message)`, isn't?

Comment (kamilmysliwiec) [reply packages/platform-express/adapters/express-adapter.ts]: would you mind investigating this https://github.com/nestjs/nest/pull/8802/files#diff-7601c6884e46e07a23aae0fea72039261b51395017f5ef74f58bb7a2519e5cb1R304 as well? @micalevisk

Comment (micalevisk) [reply packages/platform-express/adapters/express-adapter.ts]: I just tested both adapters:

```js
response.end('any string')
response.end(undefined)
response.end(null) // behave the same as above
```

```
reply.raw.end('any string')
reply.raw.end(undefined)
reply.raw.end(null) // behave the same as above
```

File: packages/platform-fastify/adapters/fastify-adapter.ts [hunk 2]
```diff
@@ -300,6 +300,10 @@ export class FastifyAdapter<
     return (response as TReply).code(statusCode);
   }
 
+  public end(response: TReply, message?: string) {
+    response.raw.end(message !== undefined ? String(message) : undefined);
```
Comment (micalevisk) [inline packages/platform-fastify/adapters/fastify-adapter.ts:304]:
```suggestion
    response.raw.end(message);
```

File: packages/platform-express/adapters/express-adapter.ts [hunk 3]
```diff
@@ -76,6 +80,10 @@ export class ExpressAdapter extends AbstractHttpAdapter {
     return this.use(handler);
   }
 
+  public isHeadersSent(response: any): boolean {
+    return response.headersSent;
```
Comment (micalevisk) [inline packages/platform-express/adapters/express-adapter.ts:84]:
```suggestion
    return !!response.headersSent;
```

in my tests, `response.headersSents` ends up being `undefined`

File: packages/platform-express/adapters/express-adapter.ts (hunk 1 above)
Comment (kamilmysliwiec) [reply packages/platform-express/adapters/express-adapter.ts]: Thank you @micalevisk!

**Pull request merged**